### DISCUSSION TITLE
Importing Airtable To NocoDB

### DISCUSSION THREAD

**Discussion by o1lab** [nocodb/nocodb#2122] [General]
### Current limitations
- Formula fields are not imported
- Rollups fields have limited support (count, min, max, avg, sum are supported)
- LongText with rich text is migrated as plain text
- All LinkToAnotherRecord are imported as many to many.
- Filters with condition groups and filters on formula fields are not imported
- Collaborator field imported as `Single line text`
- Phone number field imported as `Single line text`
- Percent field values converted as fractions (1% => 0.01)
- Duration field values converted to second's (00:00:01 => 60)
- CreateTime, LastModifiedTime are passive fields; doesn't change on edit in NocoDB
- CreatedBy, LastModifiedBy are passive fields, treated like collaborators. doesn't change on edit in NocoDB
- Button field stores URL associated directly
- Filter/Sort are not imported if they include fields not supported by NocoDB

**Comment by mahmouds12** [comment]
I am trying this feature and I get this error message:

TypeError: Cannot read properties of null (reading 'startsWith')
    at getAirtableSchema (...\nocodb\packages\nocodb\src\lib\noco\meta\api\sync\helpers\job.ts:107:21)
    
    I have the latest clone from master and connecting to a postgres db

**Reply by o1lab** [reply]
Thank you, we will look into it. Would you please DM me in [our discord](discord.nocodb.com).

**Reply by zblesk** [reply]
I'm also getting this error, with all AT bases.

**Reply by o1lab** [reply]
Are you trying our latest docker - v0.91.1 ? You can DM me in our discord.

**Reply by zblesk** [reply]
Yes, 0.91.1. I will DM you then.

**Reply by zblesk** [reply]
ℹ in case anyone else is hitting this: In my case, the issue was caused by my using the Base ID, instead of the Share Base ID which apparently is something else entirely. Thanks @o1lab for your help.

**Reply by o1lab** [reply]
How to get your shared based id is documented [here](https://docs.nocodb.com/setup-and-usages/import-airtable-to-sql-database-within-a-minute-for-free#retrieve-share-base-id--url).

**Comment by luizvaz** [comment]
Hi, I am getting this error message:


```
$ SDK initialized
$ Retrieving Airtable schema
$ Importing Tables...
$ :: [1/1] Projects
X Cannot read property 'length' of undefined
```

**Reply by luizvaz** [reply]
Thanks, guys!
Problem solved with version **0.91.1**!
**You are amazing!**

**Comment by kmataeh23** [comment]
![9C5EC4F5-0C64-4034-A9C2-342480FF2DAD](https://user-images.githubusercontent.com/70729030/181786716-d755f58c-b73f-405a-8098-367b00df8102.jpeg)

Hello I am trying to import an air table and seem to be stuck on this screen. Any suggestions? It has been like this for 20 minutes

**Reply by o1lab** [reply]
you can DM me on discord.nocodb.com

**Reply by samuraicow** [reply]
I get a "duplicate column alias" error that halts the import

**Reply by yachal-upson** [reply]
That 'duplicate column alias' is doing my head in.

**Reply by obaeyaert** [reply]
I had a column named 'alias'.
Had to rename it to fix the issue.

**Comment by kentblakely** [comment]
When I try to do the Airtable import, I get the word Importing but it seems to hang at this point. Any ideas?

**Reply by o1lab** [reply]
get in touch with @dstala on our [discord](http://discord.nocodb.com) please

**Comment by yachal-upson** [comment]
1. Have had success once before with this but keep getting 'duplicate column alias' error. 
Would love to know exactly how to hunt this down. 
2. Also every time the import process hits this error it leaves residual tables in the DB instead of deleting all on failure. 
On repeating the import process it chokes immediately on these residual tables. Have to delete project, start over. 
4. Would be amazing to be able to select specific tables to bring across to help troubleshoot.

Any thoughts?

In general, thanks to the team for the awesome self-hosted product you're bringing together. 
Can't believe Airtable refuse such an offering (or to offer reasonable pricing tiers) Their product is well done, but good as dead when it comes to anything serious. 50k rows...

**Comment by nc3d** [comment]
Any idea when formula support might be coming?
Thanks for this importer!!!

**Comment by divemasterza** [comment]
Airtable seems to have changed the URL structure for the share link i.e.:

`https://airtable.com/appAbnDph2OXXXXXXX/shrmCIxvHp2wzBQfE/tbl5gSxxxxxxxx/viwoL1h3waa6xxxxxx`

When trying to import them onto NoCoDB this is the error displayed:

![CleanShot 2023-08-31 at 11 15 40](https://github.com/nocodb/nocodb/assets/5737056/a28feb07-dc6b-4fc6-a4e1-ee17692238f3)

Any workaround/solutions?

**Reply by o1lab** [reply]
Ok, thanks for bring this to our notice. We will look into this.

**Reply by dstala** [reply]
@divemasterza can you try with just share base ID `shrmCIxvHp2wzBQfE`?
Looks like you have copied shared base URL after opening a view in shared base. NocoDB expects URL only until shared base ID. We will make changes to handle such instances.

**Reply by divemasterza** [reply]
@dstala -> not a view share it's the full URL that comes now when you share a base.
Tried also just to input the share ID but it fails.

**Reply by dstala** [reply]
You were right. Airtable supported URLs of the format `https://airtable.com/shrmCIxvHp2wzBQfE/`
But for all newly generated shared URLs, there is a mandatory app ID in URL `https://airtable.com/appAbnDph2OXXXXXXX/shrmCIxvHp2wzBQfE/`

Will be fixed in next release. Thanks @divemasterza

**Reply by rolltidehero** [reply]
Mine imports the tables and schema but not the records. I have 'Import Data' enabled. Maybe this is due to new link format?

**Reply by divemasterza** [reply]
@rolltidehero - works fine for me in 0.111.1

**Reply by ninyawee** [reply]
> Airtable seems to have changed the URL structure for the share link i.e.:
> 
> `https://airtable.com/appAbnDph2OXXXXXXX/shrmCIxvHp2wzBQfE/tbl5gSxxxxxxxx/viwoL1h3waa6xxxxxx`
> 
> When trying to import them onto NoCoDB this is the error displayed:
> 
> ![CleanShot 2023-08-31 at 11 15 40](https://private-user-images.githubusercontent.com/5737056/264598073-a28feb07-dc6b-4fc6-a4e1-ee17692238f3.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgwOTI3NTcsIm5iZiI6MTcxODA5MjQ1NywicGF0aCI6Ii81NzM3MDU2LzI2NDU5ODA3My1hMjhmZWIwNy1kYzZiLTRmYzYtYTRlMS1lZTE3NjkyMjM4ZjMuanBnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTFUMDc1NDE3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjNiMjU1NGE2ZDljMjUyZTUyOTg2Njc0OGVjMTNlYTY1ZWYyYWZjNWJiNzk2MmNjZmQzNGYzNDcwZDFlNjhjOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.nMrBcWB6TquTyNtp8zAofmMsUX6ZUl38iZYTWQRDRVI)
> 
> Any workaround/solutions?

Got similar Error to this
![image](https://github.com/nocodb/nocodb/assets/8089231/64023498-1efe-4641-b56f-26c024e2f0bb)
![image](https://github.com/nocodb/nocodb/assets/8089231/7a09921e-edfc-492c-b82d-39859770a9d2)

NocoDB is the latest `develop` branch.

I set env flag `DEBUG` to true to see log. Here is the log.
```
2024-06-11 15:08:09 [Nest] 8  - 06/11/2024, 8:08:09 AM     LOG [NestApplication] Nest application successfully started +11ms
2024-06-11 15:08:24 [Nest] 8  - 06/11/2024, 8:08:24 AM     LOG [FetchAT] AggregateError
2024-06-11 15:08:24 [Nest] 8  - 06/11/2024, 8:08:24 AM     LOG [FetchAT] Object:
2024-06-11 15:08:24 {
2024-06-11 15:08:24   "message": "Invalid Shared Base ID :: Ensure www.airtable.com/<SharedBaseID> is accessible. Refer https://bit.ly/3x0OdXI for details"
2024-06-11 15:08:24 }
2024-06-11 15:08:24 
2024-06-11 15:08:24 [Nest] 8  - 06/11/2024, 8:08:24 AM     LOG [at-import] Object:
2024-06-11 15:08:24 {
2024-06-11 15:08:24   "message": "Invalid Shared Base ID :: Ensure www.airtable.com/<SharedBaseID> is accessible. Refer https://bit.ly/3x0OdXI for details"
2024-06-11 15:08:24 }
2024-06-11 15:08:24 
2024-06-11 15:08:24 [Nest] 8  - 06/11/2024, 8:08:24 AM   ERROR [JobsEventService] ---- !! JOB FAILED !! ----
2024-06-11 15:08:24 id:1
2024-06-11 15:08:24 error:Error (Invalid Shared Base ID :: Ensure www.airtable.com/<SharedBaseID> is accessible. Refer https://bit.ly/3x0OdXI for details)
2024-06-11 15:08:24 
2024-06-11 15:08:24 stack: Error: Invalid Shared Base ID :: Ensure www.airtable.com/<SharedBaseID> is accessible. Refer https://bit.ly/3x0OdXI for details
2024-06-11 15:08:24     at AtImportProcessor.job (/usr/src/app/docker/main.js:2:1491764)
2024-06-11 15:08:24     at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
2024-06-11 15:08:24     at async QueueService.jobWrapper (/usr/src/app/docker/main.js:2:1440599)
2024-06-11 15:08:24     at async run (/usr/src/app/node_modules/p-queue/dist/index.js:163:29)
```

**Comment by sanskritkm** [comment]
Hello,

Airtable will deprecate API Key (API keys will be deprecated by the end of January 2024) and will use only [Personal access token](https://airtable.com/create/tokens)

Now if we try to import using Personal Token it will only able import table but not content in the table.

Please check.

**Comment by rolltidehero** [comment]
Has there been any fix for this? Many of us are still not able to import the Airtable data, only the table.

**Comment by croketephji** [comment]
Hi, very excited to be trying NocoDB. The limitation that I haven't been able to deal with is this one:
`All LinkToAnotherRecord are imported as many to many.`
Situation: I just imported a base from Airtable. My base has 16 tables with 18 relations between them.
Desired behavior: Linked records are shown in the appropriate column in the list view like in Airtable.
Actual behavior: All linked records are shown in the list view as "# linked items" or "No linked item". 
I think this is a result of one-to-many relations being imported as many-to -many. I cannot find a way to change the type of relation in the NocoDB UI. Is this possible? If not, can I go into the underlying DB and change the relations "under the hood"? Relinking all the records with new columns is a lot of work, especially if I have to click every linked item to find out what it's linked to...

**Reply by lukaas33** [reply]
Any fix for this? Now struggling with this in 2026

**Comment by subos2008** [comment]
I'm getting one AirTable base import including the records using a PAT. However the one that works is basically a helloworld Base with a few records. The Base I'm really interested to import which has a more data it it imports 0 records. 

I think I used an identical process to try importing each one.

No debug output that I can see.

It looks to me like there's a silent failure importing bigger/more complex tables but super simple ones with few records are working with PAT.

Happy to provide more debug if there's a place to look for it.

**Comment by luminous11111** [comment]
Hi @o1lab - When will the formula's be supported? 

Also, is there an undo function and trash bin to revert deleted bases/tables?

Thank you!

**Reply by approvingdodo** [reply]
Agree! This is the number one feature requirement preventing my team from adopting NocoDB.

**Comment by lksngy** [comment]
Hi! 

Im trying to import two different bases from Airtable to nocodb, but getting two different errors. What can I do about it?

<img width="621" alt="Screenshot 2024-05-16 at 10 26 46" src="https://github.com/nocodb/nocodb/assets/23177276/16b6114a-01db-47a1-af71-02b8a621a026">

<img width="621" alt="Screenshot 2024-05-16 at 10 36 19" src="https://github.com/nocodb/nocodb/assets/23177276/0544e53d-66da-4cf9-ab8a-e8ed1ebe8240">

**Comment by popovd91** [comment]
Anyone have suggestions on how to import an Airtable base that uses relies on a lot of formulas? I'm thinking of just rewriting those formulas on the nocodb side. Thoughts?

**Comment by timothoms** [comment]
The Airtable Import dialog has stated that Formula Import is coming soon for at least the past year. When is it actually coming?

**Comment by SelaOZ** [comment]
```/* xc */ CREATE TABLE "p7ynyyv83bqfaut"."_nc_m2m_App_catalog_App_catalog" ( "App_catalog_id" int4 NOT NULL, "App_catalog_id" int4 NOT NULL, PRIMARY KEY("App_catalog_id", "App_catalog_id")); - column "App_catalog_id" appears twice in primary key constraint```

I'm getting this error when trying to import from airtable, I don't see two identical columns in the original table. 
What can be done to make it work?

**Comment by steve-goldberg** [comment]
seems like this feature is going to be "in beta" for all eternity sadly. Any updates on this?

**Comment by approvingdodo** [comment]
No updates. I asked the developers on Discord a few weeks ago and they this is not a priority.
